patchwork-capabilities/src/main/java/net/minecraftforge/common/capabilities/CapabilityProvider.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Minecraft Forge, Patchwork Project
+ * Copyright (c) 2016-2019, 2019
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.capabilities;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import mcp.MethodsReturnNonnullByDefault;
+import net.minecraftforge.common.util.LazyOptional;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.util.math.Direction;
+
+@MethodsReturnNonnullByDefault
+@ParametersAreNonnullByDefault
+public abstract class CapabilityProvider<B> implements ICapabilityProvider
+{
+    protected final @Nonnull Class<B> baseClass;
+    protected  @Nullable CapabilityDispatcher capabilities;
+    private boolean valid = true;
+
+    protected CapabilityProvider(Class<B> baseClass)
+    {
+        this.baseClass = baseClass;
+    }
+
+    public final void gatherCapabilities() { gatherCapabilities(null); }
+
+    public void gatherCapabilities(@Nullable ICapabilityProvider parent)
+    {
+        throw new RuntimeException("No, I cant...");
```

Comment:
wat

Reply:
This class is meant to be extended by those classes which can have capabilities. This provider doesn’t have a direct instance to the actual holder, so this doesn’t work. The base implementation class however extends this method since it has an instance to the holder. Suggest a better solution :P

Reply:
abstract?